apis/pkg/v1alpha1/provider_types.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+)
+
+// +kubebuilder:object:root=true
+// +genclient
+// +genclient:nonNamespaced
+
+// Provider is the CRD type for a request to add a provider to Crossplane.
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type=='Ready')].status"
+// +kubebuilder:printcolumn:name="PACKAGE",type="string",JSONPath=".spec.package"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Cluster,categories={crossplane}
+type Provider struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ProviderSpec   `json:"spec,omitempty"`
+	Status ProviderStatus `json:"status,omitempty"`
+}
+
+// ProviderSpec specifies details about a request to install a provider to
+// Crossplane.
+type ProviderSpec struct {
+	ControllerOptions `json:",inline"`
```

Comment:
`ProviderSpec` differs from `ConfigurationSpec` in that it uses this `ControllerOptions` struct, which is also part of the `PackageRevisionSpec`. The `ConfigurationSpec` has the `ImagePullSecrets` and `ImagePullPolicy` parts of the `ControllerOptions`, but not the `ServiceAccountOptions` because the `Configuration` doesn't produce a controller that needs an SA.

We are going to need to allow configuration of the Provider's packaged controller Deployment, but I think it should be completely separate from the configuration of the unpack Job. The first step to get here would be to add `ImagePullSecrets` and `ImagePullPolicy` to `ProviderSpec` outside of the `ControllerOptions` struct. These would be the specific fields for configuring the unpack Job, just like they are in `ConfigurationSpec`.

This leaves the need to determine how to allow a user to pass configuration to the `Deployment`. We have discussed a few approaches in the past, but haven't landed on a clear direction, mostly because we don't have enough data on how folks will want to configure providers. I think we can confidently say the following:

1. Passing configuration to a Provider deployment is a desirable feature.
2. Allowing passing of configuration to a Provider deployment is a potential major security risk to anyone hosting a Crossplane. This likely means the package manager needs to be able to turn this feature on or off, similar to how it does now.
3. The default Provider deployment should be as secure as possible given the typical use case.
4. Configuration should be passed at package "install" time rather than package "build" time. If there is a configuration that is required for every deployment of a specific Provider, the package author should specify that in the documentation, and users should provide that configuration. This does require a bit of overhead, but results in a much simpler model where we don't have to worry about reconciling packaged configuration with install-time configuration. This could always be expanded in the future.
5. Configuring Provider deployments _could_ be deferred to after this initial implementation. The only known use-case right now is with IAM Roles for SA when provider-aws is installed into a Crossplane cluster running on EKS, and we are currently planning on keeping the `ClusterPackageInstall` around for a few releases, so `provider-aws` could still be installed to work in that way. Also, though it would be a degraded experience, a Provider can always be installed manually without the package manager.

With these in mind, I think I would support starting without configurability, then talking to some users and deciding the best way to support it. Alternatively, we could allow a full `DeploymentSpec` in the `Provider` install Spec, then just consider or ignore that based on how the package manager is configured, but I would prefer we bias towards not adding fields until we know how we want them to be used. The cost of adding is much less than the cost of removing.

How does all of this sound? /cc @lukeweber @negz

Reply:
> While I agree that folks could work around the issue by manually deploying the provider, I suspect it would be a better UX if they didn't have to.

yes, we definitely need to make this a good experience, but it seems that there are some barriers to doing so right now that may take time to work out in a long-term sustainable way. One other thing to mention here as a use-case is that the `imagePullPolicy` and `imagePullSecrets` also flow down to the `Deployment` right now. This allows you to use an image from a private registry for the controller `Deployment`. The uses of this are likely fairly limited right now because this is only applicable for Providers and I am not sure how many proprietary Providers are in existence. However, we could allow this to continue to flow down for now, and just document the behavior change in the future.

Reply:
Ping @turkenh. We should discuss what you see is minimum requirements for initial implementation here.

Reply:
I believe after discussion with @turkenh we're good to proceed without exposing any options to configure the provider (for now), right?

Reply:
Correct!